### ISSUE TITLE
Allow users to easily access workload endpoints

### ISSUE THREAD

**Issue by catherineluse** [rancher/dashboard#4974]
# Problem
Clicking an endpoint link for a workload in the Rancher UI usually just opens a new, blank tab. This isn't useful.

Users don't like the fact that the Ember UI gave them a shortcut to easily access an app deployed on Kubernetes for troubleshooting or development purposes, whereas the links don't work in the new UI.

# Background - Ember vs Dashboard

In the list of workloads, the endpoint column was meant to be a way to reach a workload or app using the best method available for the browser to reach it.

Ember and the Norman API provided these shortcuts by making assumptions that should not be made in Steve. For example, Ember assumed there was a one-to-one mapping between workloads and services because it automatically created a special service for each new workload. Steve doesn't do that, preferring to mimic the behavior of upstream K8s.

While the most common use case is to have one ingress point to one service and for one service point to all the Pods of one workload, K8s treats these resources as decoupled and the relationships could be many-to-many. For this reason it could be complicated to show the best URL for a workload.

Vince said Ember used this kind of logic to show the best URL:

> An ingress is almost always going to point to a service which selects (all) the pods of a single workload, and if you have a workload that has a corresponding ->service->ingress that DNS name is probably the one you want to be using to go open up that app.  So that could be worked out and just show the DNS name in that column to get you to it, as the ingress page does.
>
> Or there might a nodeport service and you can link to that, or clusterip for a http-looking port and there's the k8s service proxy, etc.

As for doing a reverse-lookup using selectors, Vince said we could load all load all workloads/services/ingresses and use things in utils/selector.js to find which things match each other, or have Steve keep what we want up-to-date in memory on the server side so it can be returned quickly.

Related issue: https://github.com/rancher/dashboard/issues/4845

# Screenshots

### Creating a Deployment with no service or ports exposed in Ember
When you create a deployment in Ember and you don't fill out any port information, you don't get an endpoint link in the workload list view or the workload detail page:
<img width="1289" alt="Screen Shot 2022-07-11 at 2 28 57 PM" src="https://user-images.githubusercontent.com/20599230/178361799-9b0ff76a-43aa-4f73-a299-9d9bd3e389bf.png">

### Creating a Deployment with a service in Ember
When you do expose your workload in the workload creation form, we have the user fill out "port" information but we don't use the word service, even though three of the four options we offer to the user are types of services:
<img width="1276" alt="Screen Shot 2022-07-11 at 2 34 35 PM" src="https://user-images.githubusercontent.com/20599230/178362883-9518fc8d-12a6-4f43-959c-f96b653f0bf6.png">

(In this workflow, we also hid the ability to add labels far at the bottom of this page, under advanced options. That's bad for app developers transitioning to the dashboard from Ember, because if you want to manually add a service in the new UI, you have to add labels to the workload, because when you create a service you will need to select the workload using the same labels.)

### Workload list view in Ember
In the screenshot below, `hello-world2` was deployed using the load balancer service type in Ember, whereas the `hello-world` was deployed in the dashboard and has no service.

The `80/tcp` link is a working link:

<img width="888" alt="Screen Shot 2022-07-11 at 2 35 35 PM" src="https://user-images.githubusercontent.com/20599230/178362970-bdae98d7-d283-4bfd-8d6b-835044209bca.png">

If you click `80/tcp`, the link goes to the load balancer endpoint (in this case an AWS load balancer because I have enabled the Amazon cloud provider). It shows that the application is working:
<img width="907" alt="Screen Shot 2022-07-11 at 2 39 23 PM" src="https://user-images.githubusercontent.com/20599230/178363235-22264f06-45f5-4221-a479-82dcbe77bf56.png">

The same working link (`80/tcp`) is on the workload detail page (this issue is mainly about replacing the functionality of this link):
<img width="627" alt="Screen Shot 2022-07-11 at 2 53 10 PM" src="https://user-images.githubusercontent.com/20599230/178365316-7e9d75b4-033b-4979-8769-7932efe2399b.png">

### Ember's automatically created services

In Ember's **Service Discovery** tab, we can see that when we deployed the app, Ember automatically created two services without telling us. In our case the new services are `hello-world2` and `hello-world2-loadbalancer`:

<img width="1377" alt="Screen Shot 2022-07-11 at 2 41 29 PM" src="https://user-images.githubusercontent.com/20599230/178364095-91cd2877-820f-47ed-b018-7fb445a4e3bd.png">

The first service is of type `ClusterIP` and it uses this selector:

```yaml
selector:
    workload.user.cattle.io/workloadselector: deployment-default-hello-world2
```

That selector matches a label that Ember automatically added to the workload.

The second automatically created service is of type `LoadBalancer`. It uses the same workload selector.
Its status shows that it has been implemented with an Amazon load balancer:

```yaml
status:
  loadBalancer:
    ingress:
    - hostname: ab58d280585654f9990b4eb7658b8601-133694235.us-west-2.elb.amazonaws.com
```

## Dashboard
When I go to a similar workload that was created in the dashboard, the masthead does not show an endpoint link if you have no service that selects it yet:
<img width="722" alt="Screen Shot 2022-07-11 at 3 00 21 PM" src="https://user-images.githubusercontent.com/20599230/178366212-07aa6087-b3a8-444e-8407-6954d2e77ef1.png">

Then I create a service of type loadbalancer. It listens on port 80 and forwards traffic to port 8080 on the container, which is where my hello world app is running:

<img width="1212" alt="Screen Shot 2022-07-11 at 3 04 27 PM" src="https://user-images.githubusercontent.com/20599230/178366683-8a0066e4-66c7-49ca-974e-f6b3edced809.png">

My workload has the label `app.kubernetes.io/name: load-balancer-example` in its spec, so I add the same label to my new service:
<img width="1174" alt="Screen Shot 2022-07-11 at 3 06 38 PM" src="https://user-images.githubusercontent.com/20599230/178366834-69d7b525-db6b-4902-809e-22398e8b727f.png">

Now if I go to the workload detail page, the masthead shows a blue endpoint link:
<img width="717" alt="Screen Shot 2022-07-11 at 3 07 38 PM" src="https://user-images.githubusercontent.com/20599230/178366986-44b08893-7fda-4b6c-8bc8-9259c0ec05f9.png">

But when clicked, this link does not work, as you can't see the application responding with "Hello Kubernetes!":
<img width="1179" alt="Screen Shot 2022-07-11 at 3 08 32 PM" src="https://user-images.githubusercontent.com/20599230/178367171-2de60f9e-90af-46f0-ad74-b47c3a86644b.png">
This is the bad UX that we mainly want to fix.

To get it to work, I create another service, this one of type ClusterIP, using the same port details and selector as my LoadBalancer service. So finally, after manually creating services that are similar to the ones that were automatically created in Ember, NOW the `80/tcp` link on the workload detail page works:
<img width="956" alt="Screen Shot 2022-07-11 at 3 12 16 PM" src="https://user-images.githubusercontent.com/20599230/178367594-61ef138b-7a4d-4f9b-b15c-7faf79baf3f0.png">

# Potential Solutions

- In the dashboard UI, we should not be showing dead links as clickable and blue. So I think if we show the endpoint data in the masthead, we should check to see if the link actually works before showing it as clickable and blue.
- We might want to add information to the workload detail page that gives users hints as to how to access it. For example, 
  - if there is a clusterIP, we could say in plain English, "This workload is exposed internally using a ClusterIP service"
  - if there is a service of type NodePort or LoadBalancer, we could say "This workload is exposed externally using a LoadBalancer service"
  - of there is no service that selects it at all, we could say, "This workload is not exposed with a service. To access it for troubleshooting or development, you could use port forwarding or create a service"
  - The workload detail page would have an "expose workload" button, or the service/ingress tabs in this PR https://github.com/rancher/dashboard/pull/6311 could have "create service" and "create ingress" buttons - but that would tie more into https://github.com/rancher/dashboard/issues/4518
- In my opinion, even if the user has no services deployed, they should still be able to have a shortcut to easily access their application. With kubectl, you can easily access your application by using a port forwarding command such as this one: 

```
kubectl port-forward deployment/hello-world 80:8080
```

The problem with that command is that it forwards port 80 on localhost, so it wouldn't work if we put it in the Rancher UI. For it to work in the UI, we would need to have it set up with a proxy so that you could access the app without a service.

I think the implementation of the endpoint link in the masthead shouldn't rely on services to work, because we can't assume that any service exists. And since this PR https://github.com/rancher/dashboard/pull/6311 already added service and ingress tabs to the workload detail page, if the user is interested to see what service endpoints exist, they can just go to that tab.

Note: It's not just the workload endpoint link that is often a dead link. In the service list view, we also have a lot of endpoint links that are just dead links.

**Comment by catherineluse** [comment]
The detail page will also have endpoints per https://github.com/rancher/dashboard/issues/5463, so this enhancement should apply to the detail page as well so the endpoints are consistent

**Comment by guyta-essence** [comment]
me to, i want the Endpoints link in workload list

**Comment by kwwii** [comment]
If possible, add a link to the endpoint so that clicking it opens the endpoint in a new window. If not possible, do not show this text as a link but normal text which has no click action

**Comment by gaktive** [comment]
Other work likely may have sorted this out.
